src/main/java/delight/nashornsandbox/internal/NashornDetection.java

```diff
@@ -0,0 +1,92 @@
+package delight.nashornsandbox.internal;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationTargetException;
+
+public class NashornDetection {
+
+    private static final Logger logger = LoggerFactory.getLogger(NashornDetection.class);
+
+    private static final Class<?> JDK_NASHORN_ScriptObjectMirror_CLASS;
+    private static final Class<?> JDK_NASHORN_NashornScriptEngineFactory_CLASS;
+    private static final Class<?> JDK_NASHORN_ClassFilter_CLASS;
+
+    private static final Class<?> STANDALONE_NASHORN_ScriptObjectMirror_CLASS;
+    private static final Class<?> STANDALONE_NASHORN_NashornScriptEngineFactory_CLASS;
+    private static final Class<?> STANDALONE_NASHORN_ClassFilter_CLASS;
+
+    static {
+        JDK_NASHORN_ScriptObjectMirror_CLASS = findClass("jdk.nashorn.api.scripting.ScriptObjectMirror", "JDK-provided Nashorn not found");
+        if (JDK_NASHORN_ScriptObjectMirror_CLASS != null) {
+            JDK_NASHORN_NashornScriptEngineFactory_CLASS = findClass("jdk.nashorn.api.scripting.NashornScriptEngineFactory", "");
+            JDK_NASHORN_ClassFilter_CLASS = findClass("jdk.nashorn.api.scripting.ClassFilter", "");
+        } else {
+            // no need to search for those and add more logs
+            JDK_NASHORN_NashornScriptEngineFactory_CLASS = null;
+            JDK_NASHORN_ClassFilter_CLASS = null;
+        }
+
+        STANDALONE_NASHORN_ScriptObjectMirror_CLASS = findClass("org.openjdk.nashorn.api.scripting.ScriptObjectMirror", "Standalone Nashorn not found");
+        if (STANDALONE_NASHORN_ScriptObjectMirror_CLASS == null) {
+            STANDALONE_NASHORN_NashornScriptEngineFactory_CLASS = findClass("org.openjdk.nashorn.api.scripting.NashornScriptEngineFactory", "");
+            STANDALONE_NASHORN_ClassFilter_CLASS = findClass("org.openjdk.nashorn.api.scripting.ClassFilter", "");
+        } else {
+            STANDALONE_NASHORN_NashornScriptEngineFactory_CLASS = null;
+            STANDALONE_NASHORN_ClassFilter_CLASS = null;
+        }
+        // TODO add a report of what was detected and what will be used
+    }
+
+    public static boolean isJDKNashornScriptObjectMirror(Object script) {
+        return JDK_NASHORN_ScriptObjectMirror_CLASS != null && JDK_NASHORN_ScriptObjectMirror_CLASS.isInstance(script);
+    }
+
+    public static boolean isStandaloneNashornScriptObjectMirror(Object script) {
+        return STANDALONE_NASHORN_ScriptObjectMirror_CLASS != null && STANDALONE_NASHORN_ScriptObjectMirror_CLASS.isInstance(script);
+    }
+
+    public static SandboxClassFilter createSandboxClassFilter() {
+        // TODO allow to force one impl?
```

Comment:
hmm interesting question - probably that would be a property on the Sandbox?

Reply:
I was thinking of 2 properties:
- prefer JDK or standalone nashorn, in case both are present
- force the use of one of those, throw an exception if not available

Also, and maybe that starts to be overkill, we could do:
- a property on the sandbox to have fine grained control, that could be useful for people that want to do it in tests
- a System property, so users can set the properties on the JVM with something like `-Ddelight-nashorn-sandbox.nashorn.provider=standalone`
The sandbox property would have precedence, fallback to the System property, then fallback to the auto detection.

Reply:
Actually I am thinking given the kind of low level nature of this, a system property maybe sufficient?

So default to standalone if present, unless there is a system property to overwrite it?